WireUI/Sources/WireUIFoundation/Main/MainCoordinator.swift

```
@@ -0,0 +1,364 @@
+/// instances and then put or remove the content view controllers into/from `viewControllers` array.
+
+@MainActor
+public final class MainCoordinator<
+
+    SplitViewController: MainSplitViewControllerProtocol,
+    TabBarController: MainTabBarControllerProtocol,
+    NewConversationBuilder: MainCoordinatorInjectingViewControllerBuilder,
+    SelfProfileBuilder: MainCoordinatorInjectingViewControllerBuilder
+
+>: NSObject, MainCoordinatorProtocol, UISplitViewControllerDelegate, UITabBarControllerDelegate where
+
+    SplitViewController.Sidebar: MainSidebarProtocol,
+    SplitViewController.ConversationList == TabBarController.ConversationList,
+    SplitViewController.Archive == TabBarController.Archive,
+    SplitViewController.Settings == TabBarController.Settings
+{
+    // swiftlint:enable opening_brace
+
+    // MARK: - Private Properties
+
+    private weak var splitViewController: SplitViewController!
+    private weak var tabBarController: TabBarController!
+
+    private let newConversationBuilder: NewConversationBuilder
+    private weak var newConversation: UIViewController?
+
+    private var selfProfileBuilder: SelfProfileBuilder
+    private weak var selfProfile: UIViewController?
+
+    private var mainSplitViewState: MainSplitViewState = .expanded
+
+    // MARK: - Private Helpers
+
+    private var sidebar: SplitViewController.Sidebar! {
+        splitViewController.sidebar
+    }
+
+    private var conversationList: TabBarController.ConversationList! {
+        switch mainSplitViewState {
+        case .collapsed: tabBarController.conversations?.conversationList
+        case .expanded: splitViewController.conversationList
+        }
+    }
+
+    private var archive: TabBarController.Archive! {
+        switch mainSplitViewState {
+        case .collapsed: tabBarController.archive
+        case .expanded: splitViewController.archive
+        }
```

are we sure it is safe to force unwrap here ? what if splitViewController archive is nil ?

---

The logic in this class should be consistent so that those cases can't happen.
I sometimes keep a force unwrap type so that a unit test could check for `nil` without failing hard.